### DISCUSSION TITLE
Tooltips, i want to try to implement, but i need assistance

### DISCUSSION THREAD

**Discussion by brunogcar** [Mikubill/sd-webui-controlnet#474] [Ideas]
one thing that i think could help users would be adding simple explanation Weight, Guidance, preprocessor settings, etc like A1111 does to CFG Scale, Seed, etc 

so im going to try to do it myself instead of requesting, but im going to need some assistance, cos i dont know python, used to know php quite well, buut its been over a decade i dont do that, nowadays i just to some javascript to my googlesheets, those ex[eriences wont help mush, most of times i cant even get a fist of the code i see here

so after some basic research,  i came up with barebones of what i think is a start to what i want to do, lets just start with the fist setting Enable Check box

following A1111 example [javascript/hints.js](https://github.com/AUTOMATIC1111/stable-diffusion-webui/commit/a96076f49ce7755ebdf3cad588b5c0d2f1f93035#diff-ddc07d50fa3b043925b1e831b1373976798d62c9f5c11fcb16c6c830bd3857cd)
```
titles = {
  "Enable": "Click on checkbox to activate ControlNet",
}


onUiUpdate(function(){
	gradioApp().querySelectorAll('span, button, select').forEach(function(span){
		tooltip = titles[span.textContent];

		if(!tooltip){
		    tooltip = titles[span.value];
		}

		if(tooltip){
			span.title = tooltip;
		}
	})

	gradioApp().querySelectorAll('select').forEach(function(select){
	    if (select.onchange != null) return;

	    select.onchange = function(){
            select.title = titles[select.value] || "";
	    }
	})
})
```
then add

`import mimetypes`

anything else needs to be imported? 
need to call  for javascript/hints.js, probably some other file called it, from files called from /modules/ui.py that i didnt follow
and
```
# this is a fix for Windows users. Without it, javascript files will be served with text/html content-type and the bowser will not show any UI
mimetypes.init()
mimetypes.add_type('application/javascript', '.js')
```
to scripts)/controlnet.py ?

then how to make it display tooltip when you mouse over at?

```
        with gr.Row():
            enabled = gr.Checkbox(label='Enable', value=False)
```

needs to use css, but probably that can be used from a1111

would appreciate if someone could show me the way, it would be a good opportunity to try to learn, baby steps, once i get it working would expand to all settings, then then try to do some other chore while you guys focus on more important matters

**Comment by missionfloyd** [comment]
All you need is hints.js

**Comment by LonicaMewinsky** [comment]
Agreed, this would really helpful. Probably see a marked reduction in issues list. There are some [caveats](https://github.com/AUTOMATIC1111/stable-diffusion-webui/issues/7534) when implementing this, though. It's possible to interfere with other extensions and/or vanilla ui.

Couple other thoughts:

You need unique string identifiers for components to target, which would mean some changes to UI functions. For example:
```
webcam_enable = ToolButton(value=camera_symbol)
```
or
```
enabled = gr.Checkbox(label='Enable', value=False)
```
would need unique identifiers (elem_id?) to target them precisely.

Secondly, and maybe this is just me, tooltips should be absolutely reliable in their description. A bad/typo'd tooltip itself can contribute to new issues. Make sure to draw from the README for precise descriptions, and maybe even open up the discussion (collab spreadsheet?) so others can contribute/edit. I'd be more than happy to help.

**Reply by brunogcar** [reply]
thanks for pointing out those caveats, it did clarify to me some stuff i wasnt sure about, as those extensions are simpler and smaller, its easier to follow what they have done

indeed, its not just to reduce issues, but try to make it clear, some of those functions to users, mouse over tool tips are usefull, specially to new users

once i get basic functionality working with enable example, will make a spreadsheet so other ppl can suggest what would be better to explain, describe  each component

my goal here is to learn, try to help a little

**Reply by LonicaMewinsky** [reply]
![hogan-hulk](https://user-images.githubusercontent.com/93007558/224422222-71966a43-61fe-487d-90f2-6da0645904e4.gif)

**Reply by brunogcar** [reply]
that cant be that easy, what kind of back magic is that?

 i just created /javascript/controlnet_hints.js from my 1st post with some changed that you committed to your repo to prevent breaking stuff and changed 

`enabled = gr.Checkbox(label='Enable', value=False)
`
to 

` enabled = gr.Checkbox(label='Enable', value=False, elem_id='Enable')
`

dont i have to declare somewhere that extra field? dont i have to define what that extra param means? dont i have to point out where to get that info from? no css needed?

i assume some of that is related to controlnet inheriting some functionality from stable-diffusion-webui being a framework behind it , but it was way easier than i expected and was trying

bear in mind its been about 13 yearrs i dont code, and was trying to do with what i remember from html and php

**Reply by LonicaMewinsky** [reply]
Looking at the js:
```
		tooltip = gif2gif_titles[span.textContent];

		if(!tooltip){
		    tooltip = gif2gif_titles[span.value];
		}

		if(!tooltip){
			for (const c of span.classList) {
				if (c in gif2gif_titles) {
					tooltip = gif2gif_titles[c];
					break;
				}
			}
		}

		if(tooltip){
			span.title = tooltip;
		}
	})
```
It'll try to match textConent, value, and members of classList. Using elem_id won't work without some changes to the js, it looks like. That said; if you have a titles entry for 'Enable,' that will work as-is, but this restricts the use of 'Enable' as a label or value anywhere else. This lack of specificity is what I was calling out as the problem. Hey, might not even be a problem?

**Comment by brunogcar** [comment]
i was only able to make it work matching label = elem_id, tried extending Enable to Enable_ControlNet and adding _ and - to the dictionary to the other elem_id, but probably will have to define those in a separated function, and that goes beyond by skills ATM

so started creating \javascript\controlnet_hints.js

```
controlnet_titles = {
    "Enable": "Click on checkbox to activate ControlNet",
    "Invert Input Color": "Invert colors if your image has white background",
    "RGB to BGR": "Convert RGB (Red, Green and Blue)to BGR (Blue, Green and Red)",
    "Low VRAM": "(Windows) (NVIDIA: Ampere) 4gb - with `--xformers` enabled, and `Low VRAM` mode ticked in the UI, goes up to 768x832, \n please use this mode if you are using 8GB GPU(s) or if you want larger batch size",
    "Guess Mode": "Non-Prompt Mode, Experimental, \n ControlNet encoder will try best to recognize the content of the input control map (preprocessor)"
}

onUiUpdate(function(){
	gradioApp().querySelectorAll('span, button, select, p').forEach(function(span){
		tooltip = controlnet_titles[span.textContent];

		if(!tooltip){
		    tooltip = controlnet_titles[span.value];
		}

		if(!tooltip){
			for (const c of span.classList) {
				if (c in controlnet_titles) {
					tooltip = controlnet_titles[c];
					break;
				}
			}
		}

		if(tooltip){
			span.title = tooltip;
		}
	})

	gradioApp().querySelectorAll('select').forEach(function(select){
	    if (select.onchange != null) return;

	    select.onchange = function(){
            select.title = controlnet_titles[select.value] || "";
	    }
	})
})
```
and changed in controlnet.py 

```
        with gr.Row():
            enabled = gr.Checkbox(label='Enable', value=False, elem_id='Enable')
            scribble_mode = gr.Checkbox(label='Invert Input Color', value=False, elem_id='Invert Input Color')
            rgbbgr_mode = gr.Checkbox(label='RGB to BGR', value=False, elem_id='RGB to BGR')
            lowvram = gr.Checkbox(label='Low VRAM', value=False, elem_id='Low VRAM')
            guess_mode = gr.Checkbox(label='Guess Mode', value=False, elem_id='Guess Mode')
```

those are just place holders and examples, but if thats good enough i will expand to try to cover everything

i see some inconsistency in labels that i will address as well 

i also see a good chance to try to make things a bit easier, to new users

how to i display that msg if the preprocessor canny is sellected

`            gr.HTML(value='<p>Recommented models for Canny Preprocessor: control_canny, t2iadapter_canny .<br ></p>')
`
probably after 

```
            if module == "canny":
                return [
                    gr.update(label="Annotator resolution", value=512, minimum=64, maximum=2048, step=1, interactive=True),
                    gr.update(label="Canny low threshold", minimum=1, maximum=255, value=100, step=1, interactive=True),
                    gr.update(label="Canny high threshold", minimum=1, maximum=255, value=200, step=1, interactive=True),
                    gr.update(visible=True)
                ]
```

my idea is to have a warning just below the preprocessor and model dropdowns
ofc would be better to have a table with compatible, recommended models for preprocessors to fetch it, instead of hardcoded text, but that is for another day

**Comment by missionfloyd** [comment]
It uses textContent (label or value, whichever's visible), not elem_id. Not an ideal way to do it, as there could be multiple controls with the same text.

Doing it by elem_id:
```python
enabled = gr.Checkbox(label='Enable', value=False, elem_id="controlnet_enable")
```
```javascript
controlnet_titles = {
    "#controlnet_enable": "Click on checkbox to activate ControlNet"
}

onUiUpdate(function(){
    for (let [key, value] of Object.entries(controlnet_titles)) {
        gradioApp().querySelector(key).title = value;
    }
})
```
> how to i display that msg if the preprocessor canny is sellected
```python
with gr.Row():
    module_messages = {
        "none": "",
        "canny": "Recommented models for Canny Preprocessor: control_canny, t2iadapter_canny"
        # ...
    }
    module_msg = gr.HTML()
    module.change(lambda msg: module_messages[msg], inputs=module, outputs=module_msg)
```
or

Add HTML somewhere
```python
module_msg = gr.HTML()
```
add messages to [`build_sliders()`](https://github.com/Mikubill/sd-webui-controlnet/blob/c9340671d6d59e5a79fc404f78f747f969f87374/scripts/controlnet.py#L328-L405)
```python
def build_sliders(module):
    if module == "canny":
        return [
            gr.update(label="Annotator resolution", value=512, minimum=64, maximum=2048, step=1, interactive=True),
            gr.update(label="Canny low threshold", minimum=1, maximum=255, value=100, step=1, interactive=True),
            gr.update(label="Canny high threshold", minimum=1, maximum=255, value=200, step=1, interactive=True),
            gr.update(visible=True),
            "Recommented models for Canny Preprocessor: control_canny, t2iadapter_canny"
        ]
```
add output to [`module.change()`](https://github.com/Mikubill/sd-webui-controlnet/blob/c9340671d6d59e5a79fc404f78f747f969f87374/scripts/controlnet.py#L415)
```python
module.change(build_sliders, inputs=[module], outputs=[processor_res, threshold_a, threshold_b, advanced, module_msg])
```
